### ISSUE TITLE
2.x: IO Scheduler ThreadWorker can be returned to the pool while still in use

### ISSUE THREAD

**Issue by jkarshin** [ReactiveX/RxJava#6477]
RxJava: 2.2.8
Java (HotSpot): 11.0.1

I'm not sure which operator is responsible for returning the worker to the pool prematurely, but I am able to reproduce the situation consistently with the following:

```java
    @Test
    public void test() throws Exception {
        AtomicBoolean shouldLoop = new AtomicBoolean();
        AtomicInteger innerRepeatCounter = new AtomicInteger();

        Completable.fromAction(() -> maybeLoop(shouldLoop.get()))
                .timeout(500, TimeUnit.MILLISECONDS, Schedulers.io())
                .andThen(Single.just(42)
                        .doOnSuccess(x -> log("Emitting value in inner Single..."))
                        .observeOn(Schedulers.io()) // <- Removing this or changing the Scheduler eliminates the problem
                        .repeatUntil(() -> innerRepeatCounter.getAndIncrement() == 2))
                .doOnError(e -> log("Suppressing exception..."))
                .onErrorResumeNext(Flowable.empty())
                .repeatUntil(() -> shouldLoop.getAndSet(true))
                .subscribeOn(Schedulers.io())
                .blockingSubscribe();
    }

    private void maybeLoop(final boolean shouldLoop) throws InterruptedException {
        try {
            if (shouldLoop) {
                log("Entering loop...");
                while (true) {
                    Thread.sleep(1000);
                }
            }
            log("Skipped loop.");
        }
        catch (InterruptedException e) {
            log("Interrupted in loop...");
            throw e;
        }
    }

    private void log(final String s) {
        System.out.println(Thread.currentThread().getName() + ": " + s);
    }
```

It looks a little contrived since I've removed as much non-Rx logic as I could, but it essentially does the following:

1) Derive a `Completable` from the `maybeLoop` method. The first time this method is called, it will do nothing and return immediately.
2) A timeout is applied to the loop in step 1. The first time through, the timeout will not be triggered since `maybeLoop` exits immediately.
3) The `andThen` creates a `Single` with an arbitrary value, uses `observeOn` to switch threads, and repeats for a total of 3 iterations.
4) Outside the `andThen`, a `doOnError` and `onErrorResumeNext` are provided to log and suppress any `TimeoutExceptions` from step 2.
5) Steps 1 through 4 are repeated. This time around, the loop in step 1 is invoked.
6) The loop is set up to loop forever. It is expected that the timeout (from step 2) will be triggered, causing the loop to be interrupted. A `TimeoutException` is expected to be propagated, bypassing the `andThen` in step 3. The exception is caught and logged in step 5, the repeat in step 6 is not triggered, and the flow completes.

Here's output for what actually happens:
```
RxCachedThreadScheduler-1: Skipped loop.
RxCachedThreadScheduler-1: Emitting value in inner Single...
RxCachedThreadScheduler-2: Emitting value in inner Single...
RxCachedThreadScheduler-2: Emitting value in inner Single...
RxCachedThreadScheduler-2: Entering loop...

// No further output and the test hangs
```

It appears that steps 1 through 4 work as expected. However, on the second iteration the timeout is never invoked and the code hangs indefinitely (due to the infinite loop).

I stuck a breakpoint inside `CompletableTimeout` on line 47 (where the `DisposeTask` is being scheduled) to investigate why the dispose was never happening. The **second** time this breakpoint was hit, I traced through the call to `Scheduler.scheduleDirect(...)` and noticed that the `Worker` created on line 198 of `Scheduler.scheduleDirect(...)` was wrapping the same thread that was currently executing (`RxCachedThreadScheduler-2`).

In other words, it appears that the `RxCachedThreadScheduler-2` thread is being returned to the thread pool while it is still being used. It is then used to try to schedule the `DisposeTask` which handles the timeout logic in step 2. Since `RxCachedThreadScheduler-2` is the current thread, it is used to execute the infinite loop. This means that it is never free to execute the `DisposeTask` which performs the timeout.

I also noticed that removing the `observeOn(...)` within the `andThen` or changing its `Scheduler` to something other than IO causes the code to terminate as expected. This might suggest that the `observeOn(...)` has something to do with the problem.

Apologies for the wall of text =). Let me know if you need any more info. Thanks!

**Comment by akarnokd** [comment]
This can happen if you have unconfined blocking code and using a trampolined resubscription operator, suchas `repeatUntil`. Quite some time ago, there was an issue of the io() scheduler creating more and more threads when it released its workers late. 
Make sure the blocking code is executed on a fresh thread via `Completable.fromAction().subscribeOn()`.

**Comment by jkarshin** [comment]
Thanks for the quick reply.

Moving the `subscribeOn()` above the `repeatUntil(...)` seems to fix the problem. I was unaware that moving the `subscribeOn()` call would have an effect on the scheduling logic; http://reactivex.io/documentation/operators/subscribeon.html seems to imply that it does not matter where `subscribeOn(...)` is called:

> ... the SubscribeOn operator designates which thread the Observable will begin operating on, no matter at what point in the chain of operators that operator is called.

Is there any downside to calling `subscribeOn(...)`as soon as possible? In other words, is `Completable.fromAction(...).subscribeOn(...)./*more operators*/` always preferable to `Completable.fromAction(...)./*more operators*/.subscribeOn(...)`?

**Comment by akarnokd** [comment]
It does matter as the subscription effect travels upstream from that point on. `repeat` calls `subscribe` again thus if a `subscribeOn` is after it, that will have no effect.

> Is there any downside to calling

It is the preferred. Put `subscribeOn` as close to the source operator as possible.
